platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterRetriever.java

```diff
@@ -145,59 +145,67 @@ public static long getRound(@NonNull final State state) {
+        // So we need to be able to handle this situation here:
+        final Bytes cert;
         try {
-            // There's code, especially in tests, that creates AddressBooks w/o any certificates/keys
-            // (because it would be time-consuming, and these tests don't use the keys anyway.)
-            // So we need to be able to handle this situation here:
-            final Bytes cert = address.getSigCert() == null
+            cert = address.getSigCert() == null
                     ? Bytes.EMPTY
                     : Bytes.wrap(address.getSigCert().getEncoded());
-            return RosterEntry.newBuilder()
-                    .nodeId(address.getNodeId().id())
-                    .weight(address.getWeight())
-                    .gossipCaCertificate(cert)
-                    .gossipEndpoint(List.of(
-                                    Pair.of(address.getHostnameExternal(), address.getPortExternal()),
-                                    Pair.of(address.getHostnameInternal(), address.getPortInternal()))
-                            .stream()
-                            .filter(pair -> pair.left() != null && !pair.left().isBlank() && pair.right() != 0)
-                            .distinct()
-                            .map(pair -> {
-                                final Matcher matcher = IP_ADDRESS_PATTERN.matcher(pair.left());
-
-                                if (!matcher.matches()) {
-                                    return ServiceEndpoint.newBuilder()
-                                            .domainName(pair.left())
-                                            .port(pair.right())
-                                            .build();
-                                }
-
-                                try {
-                                    return ServiceEndpoint.newBuilder()
-                                            .ipAddressV4(Bytes.wrap(new byte[] {
-                                                (byte) Integer.parseInt(matcher.group(1)),
-                                                (byte) Integer.parseInt(matcher.group(2)),
-                                                (byte) Integer.parseInt(matcher.group(3)),
-                                                (byte) Integer.parseInt(matcher.group(4)),
-                                            }))
-                                            .port(pair.right())
-                                            .build();
-                                } catch (NumberFormatException e) {
-                                    throw new InvalidAddressBookException(e);
-                                }
-                            })
-                            .toList())
-                    .build();
         } catch (CertificateEncodingException e) {
             throw new InvalidAddressBookException(e);
         }
+
+        final List<ServiceEndpoint> endpoints = AddressBookUtils.endpointsFor(address);
```

Comment:
This `endpointsFor` method returns endpoints in an order that is opposite to what the Roster expects. In the Roster, the external endpoint goes at index 0, followed by an optional internal endpoint. The internal endpoint may be skipped if it's identical to the external one.

The entire platform code relies on this particular order of endpoints in the Roster and will be broken if it's reversed. This is because we're going to deprecate internal endpoints from the Roster eventually and replace them with the node configuration. So we choose to use the entry at index 0 from the start to refer to an external endpoint that the platform normally uses.

It would be best to retain the existing conversion logic in the RosterRetriever class as is.

Reply:
I fully support the intent of this change per your description. And if this conversion logic must be re-written, that's okay, too.

Speaking of this particular line of code though, since this call occurs in a Roster-related class and we know that the `AddressBookUtils` would return the endpoints in an order that is opposite to the roster specification, it would be a lot more appropriate to flip the entries right here, instead of burying the flipping logic down the call stack in a completely different method. This would help ensure that the rest of the code in this class doesn't need to deal with flipping at all.

However, per a comment below in this PR, it does seem that there's an extra place which requires the flipping logic - in the RosterUtils. It would be ideal to find a way to have this flipping implemented literally once in a single place somehow, to follow the spirit of the intent of this fix.